internal/shim-sev/src/hostcall.rs

```diff
@@ -64,175 +89,509 @@ fn return_empty_option<'a>(_i: usize) -> Option<&'a mut Block> {
+
+        Some(Self {
+            block_guard: bg,
+            tls,
+        })
+    }
+
+    /// get an SNP attestation report
+    ///
+    /// See https://github.com/enarx/enarx/issues/966
+    pub fn get_attestation(
+        &mut self,
+        nonce: usize,
+        nonce_len: usize,
+        buf: usize,
+        buf_len: usize,
+    ) -> Result<[usize; 2], libc::c_int> {
+        if !snp_active() {
+            return Ok([0, 0]);
+        }
+
+        if buf == 0 {
+            return Ok([SNP_ATTESTATION_LEN_MAX, TECH]);
+        }
+
+        if buf_len > isize::MAX as usize {
+            return Err(libc::EINVAL);
+        }
+
+        if buf_len < SNP_ATTESTATION_LEN_MAX {
+            return Err(libc::EMSGSIZE);
+        }
+
+        if nonce_len != 64 {
+            return Err(libc::EINVAL);
+        }
+
+        let nonce = self.validate_slice::<u8>(nonce, nonce_len)?;
+
+        let buf = self.validate_slice_mut::<u8>(buf, buf_len)?;
+
+        let len = GHCB_EXT.get_report(1, nonce, buf).map_err(|_| libc::EIO)?;
+
+        Ok([len, TECH])
+    }
+}
+
+impl Handler for HostCall<'_> {
+    #[inline(always)]
+    fn block(&self) -> &[usize] {
+        self.block_guard.block.as_ref().unwrap()
+    }
+
+    #[inline(always)]
+    fn block_mut(&mut self) -> &mut [usize] {
+        self.block_guard.block.as_mut().unwrap()
+    }
+
+    #[inline(always)]
+    fn thread_local_storage(&mut self) -> &mut ThreadLocalStorage {
+        self.tls
+    }
+
+    fn arch_prctl(&mut self, code: c_int, addr: c_ulong) -> sallyport::Result<()> {
+        match code {
+            syscall::ARCH_SET_FS => {
+                // FIXME: check `addr` value
+                unsafe {
+                    FS::write_base(VirtAddr::new(addr));
+                }
+                eprintln!("SC> arch_prctl(ARCH_SET_FS, {:#x}) = 0", addr);
+                Ok(())
+            }
+            syscall::ARCH_GET_FS => {
+                let addr: &mut u64 = self.validate_mut(addr as _)?;
+                *addr = FS::read_base().as_u64();
+                Ok(())
+            }
+            syscall::ARCH_SET_GS => {
+                // FIXME: check `addr` value
+                unsafe {
+                    GS::write_base(VirtAddr::new(addr));
+                }
+                eprintln!("SC> arch_prctl(ARCH_SET_GS, {:#x}) = 0", addr);
+                Ok(())
+            }
+            syscall::ARCH_GET_GS => {
+                let addr: &mut u64 = self.validate_mut(addr as _)?;
+                *addr = GS::read_base().as_u64();
+                Ok(())
+            }
+            x => {
+                eprintln!("SC> arch_prctl({:#x}, {:#x}) = -EINVAL", x, addr);
+                Err(libc::EINVAL)
+            }
+        }
+    }
+
+    fn brk(&mut self, addr: Option<NonNull<c_void>>) -> sallyport::Result<NonNull<c_void>> {
+        let len;
+
+        let next_brk = *NEXT_BRK_RWLOCK.read().deref();
+        let virt_addr = next_brk;
```

Comment:
Looking at the values and usage, it seems that these variables are identical, why not just use one?

Reply:
so true .. thank you